File: src/player.py
import util
import random
import card


class Player(object):
    def __init__(self):
        self.deck = []
        self.discard = [card.Estate, card.Estate, card.Estate,
                        card.Copper, card.Copper, card.Copper, card.Copper,
                        card.Copper, card.Copper, card.Copper]
        self.hand = []
        self.score = 0

    def __str__(self):
        result = ""
        result += "Hand:\n"
        for c in self.hand:
            result += str(c) + " "
        result += "\n"
        result += "Deck: %d cards\n" % len(self.deck)
        result += "Discard: %d cards\n" % len(self.discard)
        return result

    def cleanup(self, play):
        util.move_list(play, self.discard)
        util.move_list(self.hand, self.discard)
        self.draw(5)

    def shuffle(self):
        util.move_list(self.discard, self.deck)
        random.shuffle(self.deck)

    def draw(self, count):
        if count == 0:
            return
        if not self.deck:
            if not self.discard:
                return
            self.shuffle()
        self.hand.append(self.deck.pop())
        self.draw(count - 1)

File: src/dominion/player.go
package dominion

import (
	"fmt"
	"sort"

	"github.com/golang/go/src/math/rand"
)

type Player struct {
	Actions int
	Coin    int
	Buys    int
	Deck    Cards
	Hand    Cards
	Discard Cards
	Play    Cards
}

func NewPlayer(deck []*Card) *Player {
	return &Player{Deck: deck}
}

func (p *Player) String() string {
	var result string
	result += "Deck: " + p.Deck.String() + "\n"
	result += "Hand: " + p.Hand.String() + "\n"
	result += "Play: " + p.Play.String() + "\n"
	return result
}

func (p *Player) Shuffle() {
	rand.Shuffle(len(p.Deck), func(i, j int) { p.Deck[i], p.Deck[j] = p.Deck[j], p.Deck[i] })
}

func (p *Player) Reshuffle() {
	p.Deck = append(p.Deck, p.Discard...)
	p.Discard = nil
	p.Shuffle()
}

func (p *Player) Draw(n int) {
	l := len(p.Deck)
	if l < n {
		if l > 0 {
			p.Draw(l)
		}
		if len(p.Discard) > 0 {
			p.Reshuffle()
			p.Draw(n - l)
		}
		return
	}
	p.Hand = append(p.Hand, p.Deck[:n]...)
	p.Deck = p.Deck[n:]
}

func (p *Player) Cleanup() {
	p.Hand.MoveTo(p.Discard)
	p.Play.MoveTo(p.Discard)
}

// Play p.Hand[card]
func (p *Player) PlayCards(cards ...int) {
	sort.Ints(cards)
	var hand Cards
	for i, c := range p.Hand {
		if len(cards) == 0 {
			break
		}
		if i == cards[0] {
			p.Play = append(p.Play, c)
			cards = cards[1:]
		} else {
			hand = append(hand, c)
		}
	}
	if len(cards) != 0 {
		panic(fmt.Sprintf("%d is out of range of hand (len %d)", cards[0], len(p.Hand)))
	}
	p.Hand = hand
}

func (p *Player) PlayTreasures() {
	var ts []int
	for i, c := range p.Hand {
		if c.Type.Is(Treasure) {
			ts = append(ts, i)
		}
	}
	p.PlayCards(ts...)
}

type Choice struct {
	Make        func(*Game)
	Description string
}

type Decision []*Choice
